snap/validate.go

```diff
@@ -71,6 +73,92 @@ func ValidateAlias(alias string) error {
 	return nil
 }
 
+// ValidateAppSocketName checks if a string ca be used as a name for a socket
+// (for socket activation).
+func ValidateAppSocketName(name string) error {
+	valid := validSnapName.MatchString(name)
+	if !valid {
+		return fmt.Errorf("invalid socket name: %q", name)
+	}
+	return nil
+}
+
+// ValidateAppSocketListenAddress checks that the value of socket addresses.
+func ValidateAppSocketListenAddress(socket *SocketInfo, fieldName string, address string) error {
+	if address == "" {
+		return fmt.Errorf("socket %q must define %q", socket.Name, fieldName)
+	}
+
+	switch address[0] {
+	case '/', '$':
+		return validateAppSocketListenAddressPath(socket, fieldName, address)
+	case '@':
+		return validateAppSocketListenAddressAbstractSocket(socket, fieldName, address)
+	default:
+		return validateAppSocketListenAddressNetAddress(socket, fieldName, address)
+	}
+}
+
+func validateAppSocketListenAddressPath(socket *SocketInfo, fieldName string, path string) error {
+	if path != filepath.Clean(path) {
+		return fmt.Errorf(
+			`socket %q has invalid %q: paths must not include "." or ".."`, socket.Name, fieldName)
+	}
+
+	if !(strings.HasPrefix(path, "$SNAP_DATA/") || strings.HasPrefix(path, "$SNAP_COMMON/")) {
+		return fmt.Errorf(
+			"socket %q has invalid %q: only $SNAP_DATA and $SNAP_COMMON prefixes are allowed", socket.Name, fieldName)
+	}
+
+	return nil
+}
+
+func validateAppSocketListenAddressAbstractSocket(socket *SocketInfo, fieldName string, path string) error {
+	prefix := "@snap." + socket.App.Snap.Name()
+	if !strings.HasPrefix(path, prefix) {
+		return fmt.Errorf("socket %q path for %q must be prefixed with %q", socket.Name, fieldName, prefix)
+	}
+	return nil
+}
+
+func validateAppSocketListenAddressNetAddress(socket *SocketInfo, fieldName string, address string) error {
```

Comment:
I have a problem with the whole validation code as written: it takes me more time to read the function name, and then find it in the sea of other functions that are all similarly named and often a one-line check, than it takes me to read the function itself. Compounding things the functions all take a bunch of arguments, all slightly different, and only some of them are involved in the validation itself; the rest are for generating the error message.

What's the reasoning for doing it this way? It's really hard to follow.

I can understand the need of having the validators separate, but surely it'd be easier and cleaner to have, say, `isValidNetAddress(string) bool`?

Reply:
The naming is following the style of other validators in this module. I agree names are quite verbose, and in this specific case the validator is pretty trivial, but `isValidNetAddress` is not a good name, since it'd need to be `isValidNetAddressForListenAddress` or something similar (as it's specific to socket activation).

I'm open to better names and I'm fine with inlining all the sub-validators for ListenAddress if that's easier to follow

Reply:
at least for me it would be. I'll ask somebody else to chip in here.

Reply:
@Chipaca is right.. some of these names are a bit over the top. That said, @albertodonato is also right. We have a local convention for these names which would be nice to respect.

Here is a suggested way forward:

| old | new |
| --- | --- |
| ValidateAppSocketName | validateSocketName |
| ValidateAppSocketListenAddress | validateSocketAddr |
| validateAppSocketListenAddressPath | validateSocketAddrPath |
| validateAppSocketListenAddressAbstractSocket | validateSocketAddrAbstract |
| validateAppSocketListenAddressNetAddress | validateSocketAddrNet |
| validateAppSocketListenAddressNetAddressAddress | validateSocketAddrNetHost |
| validateAppSocketListenAddressNetAddressPort | validateSocketAddrNetPort |

Reply:
NetHost and NetPort might indeed be inlined, btw.

Reply:
Renamed as suggested. I didn't inline `validateSocketAddrNetHost` and `validateSocketAddrNetPort`, since the latter is used in two places, and I'd leave the first one for symmetry.